python/cudf/cudf/core/indexed_frame.py

```diff
@@ -761,6 +762,42 @@ def add_suffix(self, suffix):
                 Use `Series.add_suffix` or `DataFrame.add_suffix`"
         )
 
+    @annotate("APPLY", color="purple", domain="cudf_python")
+    def _apply(self, func, kernel_getter, *args, **kwargs):
+        """Apply `func` across the rows of the frame."""
+        if kwargs:
+            raise ValueError("UDFs using **kwargs are not yet supported.")
+
+        kernel, retty = _compile_or_get(
+            self, func, args, kernel_getter=kernel_getter
+        )
```

Comment:
Should we catch the numba error when this fails and present the user with something more friendly (that doesn't require them to know that we JIT-compile with numba)? We could use a try-except with a `raise ... from e` so that knowledgeable users could see the numba error if they wanted.

Reply:
I would be +1 on this but I am a little unsure what a good singular error message would be. There's a lot of different things that can go wrong when someone uses this, such as signature/arg mismatches, unsupported dtypes, or unrecognized functions being used inside the udf, and numba is fairly good at indicating approximately what went wrong out of those things and often how to fix it. We could bookend the possible numba errors with a friendlier set of error messages that derive from the numba errors, but I think if we only had one general error I am not sure what it would say other than generally "something went wrong" which may hide the helpful stuff numba actually has to say about it.

Reply:
That's why I'm suggesting a `raise ... from`. If you haven't seen it, it looks like this:
```suggestion
        try:
            kernel, retty = _compile_or_get(
                self, func, args, kernel_getter=kernel_getter
            )
        except Exception as e:
            # Or maybe our error message is more specific and explicitly says "failed to compile"
            raise ValueError("Your function could not be successfully applied") from e
```
which will result in something like this:
```
Traceback (most recent call last):
...
ValueError: Your function could not be successfully applied

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
...
TypingError: ...
```
That way users see both the numba error and our wrapper indicating that it's a failure to apply the function.